funnel-udf/src/main.rs

```diff
@@ -88,149 +97,156 @@ fn calculate_funnel_from_user_events(
 
         for (timestamp, events_with_same_timestamp) in &filtered_events {
             let events_with_same_timestamp: Vec<_> = events_with_same_timestamp.collect();
-            entered_timestamp[0] = EnteredTimestamp {
+            vars.entered_timestamp[0] = EnteredTimestamp {
                 timestamp,
                 timings: vec![],
                 uuids: vec![],
             };
 
             if events_with_same_timestamp.len() == 1 {
-                if !process_event(
+                if !self.process_event(
+                    args,
+                    &mut vars,
                     &events_with_same_timestamp[0],
-                    &mut entered_timestamp,
-                    &prop_val,
-                    &mut event_uuids,
-                    conversion_window_limit_seconds,
-                    funnel_order_type,
-                    &mut max_step,
-                    breakdown_step,
-                    &mut results,
+                    prop_val,
+                    false
                 ) {
-                    add_max_step = false;
-                    break;
+                    return;
                 }
-            } else {
-                // Handle permutations for events with the same timestamp
-                let mut entered_timestamps: Vec<_> = vec![];
-                for perm in events_with_same_timestamp.iter().permutations(events_with_same_timestamp.len()) {
-                    entered_timestamps.push(entered_timestamp.clone());
-                    for event in perm {
-                        if !process_event(
-                            &event,
-                            &mut entered_timestamps.last_mut().unwrap(),
-                            &prop_val,
-                            &mut event_uuids,
-                            conversion_window_limit_seconds,
-                            funnel_order_type,
-                            &mut max_step,
-                            breakdown_step,
-                            &mut results,
-                        ) {
-                            add_max_step = false;
-                            break;
-                        }
+            } else if events_with_same_timestamp.iter().map(|x| &x.steps).all_equal() {
+                // Deal with the most common case where they are all the same event (order doesn't matter)
+                for event in events_with_same_timestamp {
```

Comment:
might have misunderstood the logic, but if it's the exact same event at same timestamp, why do we need to process all of them and not just one? for cases where we need to check number of times funnel is completed?

Reply:
The original funnels code allows for multiple events at the same timestamp to count in a funnel.

So if a funnel asks for events A,B,B,C and two B events come in with the same timestamp, we'd want to consider the funnel completed.